test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -8011,6 +8011,223 @@ BOOST_AUTO_TEST_CASE(struct_named_constructor)
+	    pragma experimental ABIEncoderV2;
+		contract C {
+			struct S { uint256 a; uint256 b; }
+			function f(S[] calldata s) external pure returns (uint256 l, uint256 a, uint256 b, uint256 c, uint256 d) {
+			    S[] memory m = s;
+			    l = m.length;
+			    a = m[0].a;
+			    b = m[0].b;
+			    c = m[1].a;
+			    d = m[1].b;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+
+	ABI_CHECK(callContractFunction("f((uint256,uint256)[])", encodeArgs(u256(0x20), u256(2), u256(1), u256(2), u256(3), u256(4))), encodeArgs(u256(2), u256(1), u256(2), u256(3), u256(4)));
+}
+
+
+BOOST_AUTO_TEST_CASE(calldata_struct_to_memory)
+{
+	char const* sourceCode = R"(
+	    pragma experimental ABIEncoderV2;
+		contract C {
+			struct S { uint256 a; uint256 b; }
+			function f(S calldata s) external pure returns (uint256, uint256) {
+			    S memory m = s;
+			    return (m.a, m.b);
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+
+	ABI_CHECK(callContractFunction("f((uint256,uint256))", encodeArgs(u256(42), u256(23))), encodeArgs(u256(42), u256(23)));
+}
+
+BOOST_AUTO_TEST_CASE(nested_calldata_struct)
+{
+	char const* sourceCode = R"(
+	    pragma experimental ABIEncoderV2;
+		contract C {
+			struct S1 { uint256 a; uint256 b; }
+			struct S2 { uint256 a; uint256 b; S1 s; uint256 c; }
+			function f(S2 calldata s) external pure returns (uint256 a, uint256 b, uint256 sa, uint256 sb, uint256 c) {
+			    return (s.a, s.b, s.s.a, s.s.b, s.c);
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+
+	ABI_CHECK(callContractFunction("f((uint256,uint256,(uint256,uint256),uint256))", encodeArgs(u256(1), u256(2), u256(3), u256(4), u256(5))), encodeArgs(u256(1), u256(2), u256(3), u256(4), u256(5)));
+}
+
+BOOST_AUTO_TEST_CASE(nested_calldata_struct_to_memory)
+{
+	char const* sourceCode = R"(
+	    pragma experimental ABIEncoderV2;
+		contract C {
+			struct S1 { uint256 a; uint256 b; }
+			struct S2 { uint256 a; uint256 b; S1 s; uint256 c; }
+			function f(S2 calldata s) external pure returns (uint256 a, uint256 b, uint256 sa, uint256 sb, uint256 c) {
+			    S2 memory m = s;
+			    return (m.a, m.b, m.s.a, m.s.b, m.c);
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+
+	ABI_CHECK(callContractFunction("f((uint256,uint256,(uint256,uint256),uint256))", encodeArgs(u256(1), u256(2), u256(3), u256(4), u256(5))), encodeArgs(u256(1), u256(2), u256(3), u256(4), u256(5)));
+}
+
+BOOST_AUTO_TEST_CASE(calldata_struct_short)
+{
+	char const* sourceCode = R"(
+	    pragma experimental ABIEncoderV2;
+		contract C {
+			struct S { uint256 a; uint256 b; }
+			function f(S calldata) external pure returns (uint256) {
+			    return msg.data.length;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+
+    // double check that the valid case goes through
+	ABI_CHECK(callContractFunction("f((uint256,uint256))", u256(1), u256(2)), encodeArgs(0x44));
+
+	ABI_CHECK(callContractFunctionNoEncoding("f((uint256,uint256))", bytes(63,0)), encodeArgs());
```

Comment:
@chriseth Sorry, I force pushed your comment away - but this is missing one byte, that's why it should revert.